src/components/lib/crt/crt.h

```diff
@@ -215,4 +219,107 @@ int crt_chkpt_create(struct crt_chkpt *chkpt, struct crt_comp *c);
 int crt_chkpt_restore(struct crt_chkpt *chkpt, struct crt_comp *c);
 
 
+/*
+ * - The VAS structure will be an array of pointers to `struct
+ *   crt_comp`s, and "reserved" bits. The reserved bits are used to
+ *   notate which names to *split* off, and the `crt_comp` is which
+ *   component has been `alloc_in`ed at that name. Each NS maintains a
+ *   pointer to its parent (it split off from).
+ *
+ * - We also need a similar structure like this to track the MPK
+ *   namespace and the ASID namespace.
+ */
+
+/* for 32 bit:
+ * name sz = 2^22
+ * 2^32 / 2^22 = 2^10 = 1024 names
+ * 2^10 / 2 to ensure the array fits into a page = 512 names
+ * #define CRT_VAS_NAME_SZ (1 << 22)
+ * #define CRT_VAS_NUM_NAMES 512
+ * #define CRT_MPK_NUM_NAMES 16
+ * #define CRT_ASID_NUM_NAMES 1024
+ * 
+ * for 64 bit:
+ * name sz = 2^39 
+ * 2^48 / 2^39 = 2^9 = 512 names
+ * 2^9 / 2 to ensure the array fites into a page = 256 names
+ * also: more ASIDs available in 64 bit
+ */
+
+#define CRT_VAS_NAME_SZ (1ULL << 39)
+#define CRT_VAS_NUM_NAMES 256
+#define CRT_MPK_NUM_NAMES 15
+#define CRT_ASID_NUM_NAMES 4096
+
+/* is reserved the right verbiage here? */
+struct crt_vas_name {
+	/* can change this to be just the crt comp and use the LSB for the reserved/allocated stuff */
+	u32_t reserved  : 1;
+	u32_t allocated : 1;
+	u32_t aliased   : 1;
+	struct crt_comp *comp;
+};
+
+struct crt_asid_mpk_name {
+	u32_t reserved  : 1;
+	u32_t allocated : 1;
+	/* id is index into the name array */
+};
+
+struct crt_ns_vas {
+	pgtblcap_t top_lvl_pgtbl;
+	struct crt_vas_name names[CRT_VAS_NUM_NAMES];
```

Comment:
Above, you're checking if `->names != NULL`, but since this is statically allocated, it will *always* be true. Remove the checks above?